### PR TITLE
fix: Create only one instance of `StatusChatInput`

### PR DESCRIPTION
**Pull request by igor-sirotin**
Fixes https://github.com/status-im/status-desktop/issues/10436

## Changes description

* Remove #10379 hotfix
* Moved `StatusChatInput` to `ChatColumnView`, so that it's only created once
* Added storage to `chat_content/input_area` to preserve these properties for each chat:
    * text
    * reply message id (new feature 🤘 )
    * attached images (new feature 🤘)
* Create popups, dialogs and menus dynamically

## Performance improvement

Test description:
1. Launch the app
2. Login with `status-shared-account`
3. Land on status community (as selected on previous launch)

| Description | Commit | RAM usage, Gb |
|-|-|-|
| Before changes, with #10379 hotfix | ee9db6a3024780c8fcfa4f7687af18b800f4ebd1 | 0.83 Gb |
| Before changes, without hotfix | 5ca6e7de93b18ef79a1f6f974bbc983e256da999 | 1.52 Gb |
| After changes | e45734320221161cf3700ea9a4307c1af41046ab | 0.80 Gb |

## Affected areas

chat

## Screenshot

Input preserved state demo:

https://github.com/status-im/status-desktop/assets/25482501/31ce52fa-df5f-4333-8d0d-6a613ac3f812

### CODE REVIEW

File: src/app/modules/main/chat_section/chat_content/input_area/model.nim
Comment (igor-sirotin) [inline src/app/modules/main/chat_section/chat_content/input_area/model.nim:1]: This was simply not used

File: ui/app/AppLayouts/Chat/views/ChatView.qml [hunk 1]
```diff
@@ -89,9 +89,6 @@ StatusSectionLayout {
         onOpenStickerPackPopup: {
             Global.openPopup(statusStickerPackClickPopup, {packId: stickerPackId, store: root.stickersPopup.store} )
         }
-        onOpenAppSearch: {
-            root.openAppSearch();
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/views/ChatView.qml:93]: Not needed any longer?

**Comment by jrainville** [comment]
> @igor-sirotin potentially fixing this too? #10885

Yeah I saw the code change and immediately thought that it would probably fix it

File: ui/app/AppLayouts/Chat/views/ChatView.qml (hunk 1 above)
Comment (igor-sirotin) [reply ui/app/AppLayouts/Chat/views/ChatView.qml:93]: Well it just wasn't ever called from `ChatContenView` 🙂

**Review by jrainville — COMMENTED** [review]
Great job. 

Some small comments only.

File: src/app/modules/main/chat_section/chat_content/input_area/model.nim
[the remark this answers is not in the document]
Comment (igor-sirotin) [reply src/app/modules/main/chat_section/chat_content/input_area/model.nim:1]: 🤷

File: ui/app/AppLayouts/Chat/views/ChatColumnView.qml [hunk 2]
```diff
@@ -107,46 +201,160 @@ Item {
...
+
+        Item {
+            Layout.fillWidth: true
+            Layout.fillHeight: true
+
+            Repeater {
+                id: chatRepeater
+                model: parentModule && parentModule.model
+
+                ChatContentView {
```
[the remark this answers is not in the document]
Comment (igor-sirotin) [reply ui/app/AppLayouts/Chat/views/ChatColumnView.qml:217]: There's a `Loader` inside the `ChatContentView`, which is controlled with `chatMessagesLoader.active: model.loaderActive`. The one that makes caching 5 chats.

File: ui/app/AppLayouts/Chat/views/ChatColumnView.qml [hunk 3]
```diff
@@ -107,46 +201,160 @@ Item {
...
+
+            onReplyMessageIdChanged: {
+                d.activeChatContentModule.inputAreaModule.preservedProperties.replyMessageId = replyMessageId
+            }
+
+            onFileUrlsAndSourcesChanged: {
+                d.activeChatContentModule.inputAreaModule.preservedProperties.fileUrlsAndSourcesJson = JSON.stringify(chatInput.fileUrlsAndSources)
+            }
+
+            // TODO: A lot of chatContentModule.getMyChatId is used here. Isn't it just `root.activeChatId`?
```
[the remark this answers is not in the document]
Comment (igor-sirotin) [reply ui/app/AppLayouts/Chat/views/ChatColumnView.qml:287]: I will remove it, as it won't give any benefits. Only cleanup and potential bugs :D

**Comment by igor-sirotin** [comment]
> @igor-sirotin potentially fixing this too? #10885

Tested. Unfortunately, it doesn't fix the issue 🙁

**Review by jrainville — APPROVED** [review]

**Review by osmaczko — APPROVED** [review]
Good job overall. While it's not my preference to store QML component states in the backend, I understand the constraints that led to this approach given the complexity of the `StatusChatInput` component. I'm okay with proceeding as is.

**Comment by osmaczko** [comment]
If these changes were a milk carton, they'd have 'Missing: QA' on the side :laughing:

**Comment by caybro** [comment]
@igor-sirotin the `e2e` test failure looks legit and will need fixing

**Comment by igor-sirotin** [comment]
> While it's not my preference to store QML component states in the backend

@osmaczko I agree and tried to implement this in QML at first. But the logic was quite confusing, I would have to create a `Map` holding info for each chat. And having this in backend, we have an ability later to keep these message drafts in database, or even sync between devices (ofc I'm not sure if it's ever needed 😄).

File: ui/app/AppLayouts/Chat/views/ChatColumnView.qml [hunk 4]
```diff
@@ -98,6 +99,99 @@ Item {
...
+            chatInput.textInput.cursorPosition = chatInput.textInput.length
+
+            d.restoreInputReply()
+            d.restoreInputAttachments()
+        }
+    }
+
+    Component.onCompleted: {
+        d.fixUsersStore()
+        d.fixMessageStore()
```
[the remark this answers is not in the document]
Comment (igor-sirotin) [reply ui/app/AppLayouts/Chat/views/ChatColumnView.qml]: dropped with usage of `Qt.callLater`

**Comment by igor-sirotin** [comment]
@anastasiyaig indeed, this would probably be good to test.
Can you please pick it up when there's time? 🙂 

#### What to test:
1. Chat input area basic functionality:
    - sending messages
    - opening popups, their position (gif, emoji, stickers, text styling)
    - attaching images, removing attached images
2. Input area preserves state for each chat when switching between chats:
     - text
     - images
     - reply

**Comment by anastasiyaig** [comment]
@igor-sirotin sure, it will be tested

**Comment by glitchminer** [comment]
Hi @igor-sirotin , QA results below. One query came up - is there still a limit on how many channels messages will persist in?

---
1. **OPEN** When gif selector is open it does not toggle closed when button is clicked again

https://github.com/status-im/status-desktop/assets/50769329/013c6747-2b4c-4b82-ad5b-102e627d382d

---
2. **OPEN** Alignment of added sticker packs is different to 0.12.0
image

---
Summary:
3 users + community with 10 channels

- Last open channel returned to on restart
- Send message, image, reply
- Search for emoji
- Enable Tenor, search for gif, send
- Add stickers and send
- Text formatting popup opens when message is highlighted
- Images preserved when switching channels
- Reply preserved when switching channels
- Reply with images preserved when switching channels
- Adding 5 images, removed 2, added 1 - preserved and sent OK.

**Comment by igor-sirotin** [comment]
@glitchminer thanks for testing! Fixed both points 👌 

As for a limit on channels persisted:
That's true, but only for chats themselves. So that last 5 chats would load fast when switching. But the input area preserves draft messages for all chats. Because that's not that much information comparing to chat components.

**Comment by glitchminer** [comment]
Thanks @igor-sirotin , those 2 are resolved now. Also sorry, I did just find one more regression just now related to the gif and sticker popups.

- Open stickers
- Open gifs
- Close gifs
- Stickers is still open

https://github.com/status-im/status-desktop/assets/50769329/e0de1222-33ef-4afd-a3f1-28ed1a0cc2f4

**Comment by igor-sirotin** [comment]
> * Open stickers
> * Open gifs
> * Close gifs
> * Stickers is still open

@glitchminer indeed, thanks for that!

I've fixed and simplified the logic for opening/closing emoji, gif and stickers popup. Should be fine by now.

Also, now clicking on anything inside the input area (text area, buttons, whatever) will close the popup as well. This changes the behaviour to discord-style. Previously it was possible to type the message in and still see the emoji/gif/stickers popup.

https://github.com/status-im/status-desktop/assets/25482501/b8a56d15-4ead-433c-b2b3-411832486dc4

**Comment by glitchminer** [comment]
Great, thanks @igor-sirotin. That's done now.

**Pull request merged**